frame/bridge/relayer-game/src/lib.rs

```diff
@@ -0,0 +1,445 @@
+//! # Relayer Game Module
+//!
+//! ## Assumption
+//! 1. At least **one** honest relayer
+//! 2. Each proposal's header hash is unique at a certain block height
+//!
+//!
+//! ## Flow
+//! 1. Request the header in target chain's relay module,
+//! weather the header is existed or not you should pay some fees
+//! 2. If not, target chain's relay module will ask for a proposal here
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![feature(drain_filter)]
+
+mod types {
+	// --- darwinia ---
+	use crate::*;
+
+	pub type AccountId<T> = <T as frame_system::Trait>::AccountId;
+	pub type BlockNumber<T> = <T as frame_system::Trait>::BlockNumber;
+	pub type RingBalance<T, I> = <RingCurrency<T, I> as Currency<AccountId<T>>>::Balance;
+
+	pub type TcBlockNumber<T, I> = <Tc<T, I> as Relayable>::TcBlockNumber;
+	pub type TcHeaderHash<T, I> = <Tc<T, I> as Relayable>::TcHeaderHash;
+
+	pub type GameId<TcBlockNumber> = TcBlockNumber;
+	pub type ProposalId<TcBlockNumber, TcHeaderHash> = (TcBlockNumber, TcHeaderHash);
+
+	type RingCurrency<T, I> = <T as Trait<I>>::RingCurrency;
+
+	type Tc<T, I> = <T as Trait<I>>::TargetChain;
+}
+
+// --- crates ---
+use codec::{Decode, Encode};
+// --- substrate ---
+use frame_support::{
+	debug::error, decl_error, decl_event, decl_module, decl_storage, ensure, traits::Currency,
+};
+use frame_system::{self as system, ensure_signed};
+use sp_runtime::{traits::Convert, DispatchError, RuntimeDebug};
+#[cfg(not(feature = "std"))]
+use sp_std::borrow::ToOwned;
+use sp_std::prelude::*;
+// --- darwinia ---
+use darwinia_support::{balance::lock::*, relay::*};
+use types::*;
+
+pub trait Trait<I: Instance = DefaultInstance>: frame_system::Trait {
+	type Event: From<Event<Self, I>> + Into<<Self as frame_system::Trait>::Event>;
+
+	/// The currency use for bond
+	type RingCurrency: LockableCurrency<Self::AccountId, Moment = Self::BlockNumber>;
+
+	/// A regulator to adjust relay args for a specific chain
+	type RelayerGameAdjustor: AdjustableRelayerGame<
+		Balance = RingBalance<Self, I>,
+		Moment = Self::BlockNumber,
+		TcBlockNumber = TcBlockNumber<Self, I>,
+		Sampler = Self::Sampler,
+	>;
+	type Sampler: Convert<Round, Vec<TcBlockNumber<Self, I>>> + Convert<u32, Round>;
+
+	/// The target chain's relay module's API
+	type TargetChain: Relayable;
+}
+
+decl_event! {
+	pub enum Event<T, I: Instance = DefaultInstance>
+	where
+		AccountId = AccountId<T>,
+		BlockNumber = BlockNumber<T>,
+	{
+		/// TODO
+		TODO(AccountId, BlockNumber),
+	}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait<I>, I: Instance> {
+		/// Proposal - ALREADY EXISTED
+		ProposalAE,
+		/// Target Header - ALREADY EXISTED
+		TargetHeaderAE,
+
+		/// Round - MISMATCHED
+		RoundMis,
+
+		/// Challenge - NOT HAPPENED
+		ChallengeNH,
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait<I>, I: Instance = DefaultInstance> as DarwiniaRelayerGame {
+		/// Each target chain's header relay can open a game
+		pub Games
```

Comment:
The return type is Vec<Proposal> of a game,  the storage name maybe `Game` not `Games`

Reply:
Or maybe `Proposals`, such that the term follows the `Samples`.

Reply:
Good catch